slo_generator/backends/cloud_monitoring_mql.py

```diff
@@ -0,0 +1,230 @@
+            distribution[i] = {
+                # 'upper_bound': upper_bound,
+                # 'bucket_count': bucket_count,
+                'count_sum': count_sum
+            }
+        LOGGER.debug(pprint.pformat(distribution))
+
+        if len(distribution) - 1 < threshold_bucket:
+            # maximum measured metric is below the cut after bucket number
+            lower_events_count: int = valid_events_count
+            upper_events_count: int = 0
+        else:
+            lower_events_count: int = distribution[threshold_bucket][
+                'count_sum']
+            upper_events_count: int = valid_events_count - lower_events_count
+
+        if good_below_threshold:
+            good_event_count: int = lower_events_count
+            bad_event_count: int = upper_events_count
+        else:
+            good_event_count: int = upper_events_count
+            bad_event_count: int = lower_events_count
+
+        return good_event_count, bad_event_count
+
+    def exponential_distribution_cut(self, *args, **kwargs) -> tuple[int, int]:
+        """Alias for `distribution_cut` method to allow for backwards
+        compatibility.
+        """
+        warnings.warn(
+            'exponential_distribution_cut will be deprecated in version 2.0, '
+            'please use distribution_cut instead', FutureWarning)
+        return self.distribution_cut(*args, **kwargs)
+
+    def query_sli(self, timestamp: int, window: int, slo_config: dict) -> float:
+        """Query SLI value from a given MQL query.
+
+        Args:
+            timestamp (int): UNIX timestamp.
+            window (int): Window (in seconds).
+            slo_config (dict): SLO configuration.
+
+        Returns:
+            float: SLI value.
+        """
+        # FIXME This method returns 2.0 for the 12-hour, 7-day and 28-day burn
+        #  rates of the GAE App Availability SLO computed with the `ratio` MQL
+        #  function. Not sure what I missed there.
+        measurement: dict = slo_config['spec']['service_level_indicator']
+        query: str = measurement['query']
+        series: list[TimeSeriesData] = self.query(query=query, window=window)
+        sli_value: float = series[0].point_data[0].values[0].double_value
+        LOGGER.debug(f"SLI value: {sli_value}")
+        return sli_value
+
+    def query(self, query: str, window: int) -> list[TimeSeriesData]:
+        """Query timeseries from Cloud Monitoring using MQL.
+
+        Args:
+            query (str): MQL query.
+            window (int): Window size (in seconds).
+
+        Returns:
+            list: List of timeseries objects.
+        """
+        # Enrich query to aggregate and reduce the time series over the
+        # desired window.
+        # FIXME Do this properly in a private helper function. Draw inspiration
+        #  from `_fmt_query` in the Prometheus backend.
+        query += f"| group_by [] | every {window}s | within {window}s"
```

Comment:
My biggest concern is this line. Is it the right way to replace the aggregation and reduction performed in the Cloud Monitoring backend? I have not found anything similar in MQL (yet).